scylla.go

```diff
@@ -331,3 +394,259 @@ func (p *scyllaConnPicker) randomConn() *Conn {
 	}
 	return nil
 }
+
+type scyllaDialerExt struct {
+	net.Dialer
+}
+
+var _ DialerExt = (*scyllaDialerExt)(nil)
+
+func wrapScyllaDialerExt(d *net.Dialer) *scyllaDialerExt {
+	return &scyllaDialerExt{*d}
+}
+
+func (spd *scyllaDialerExt) DialContextWithSourcePort(ctx context.Context, sourcePort uint16, network, addr string) (net.Conn, error) {
+	localAddr, err := net.ResolveTCPAddr(network, fmt.Sprintf(":%d", sourcePort))
+	if err != nil {
+		return nil, err
+	}
+
+	// Make sure that we are copying the dialer struct, not just a pointer
+	var portDialer net.Dialer = spd.Dialer
+	portDialer.LocalAddr = localAddr
+
+	return portDialer.DialContext(ctx, network, addr)
+}
+
+type scyllaShardAwarePortDialer struct {
+	innerDialer DialerExt
+
+	scConnParams *scyllaConnParams
+}
+
+func wrapScyllaShardAwarePortDialer(innerDialer Dialer, scConnParams *scyllaConnParams) Dialer {
+	innerDialerExt, isExt := innerDialer.(DialerExt)
+	if !isExt {
+		return innerDialer
+	}
+
+	if !scConnParams.usesShardAwarePort() {
+		return innerDialer
+	}
+
+	return &scyllaShardAwarePortDialer{
+		innerDialer:  innerDialerExt,
+		scConnParams: scConnParams,
+	}
+}
+
+func (ssapd *scyllaShardAwarePortDialer) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	for {
+		sourcePort := ssapd.scConnParams.sourcePort
+		if gocqlDebug {
+			Logger.Printf("scylla: connecting from port %d", sourcePort)
+		}
+		conn, err := ssapd.innerDialer.DialContextWithSourcePort(ctx, sourcePort, network, ssapd.scConnParams.targetAddress)
+		if err != nil {
+			if opErr, isOpErr := err.(*net.OpError); isOpErr {
+				if strings.Contains(opErr.Error(), "address already in use") {
+					// if we failed to bind, then probably it was due to the port being
+					// already allocated, so we can retry immediately with another port
+					if ssapd.scConnParams.updateToNextPort() {
+						continue
+					} else {
+						break
+					}
+				}
+			}
+		}
+		return conn, err
+	}
+
+	return nil, errors.New("scylla: could not allocate a free source port for connection")
+}
+
+const (
+	// the minimum port number that can be chosen for port-based load balancing
+	scyllaPortBasedBalancingMin = 0x8000
+
+	// the maximum port number that can be chosen for port-based load balancing
+	scyllaPortBasedBalancingMax = 0xFFFF
+)
+
+func (pool *hostConnPool) makeScyllaConnParams() *scyllaConnParams {
```

Comment:
This belongs to the `connectionpool.go` so why would you put it here?

Reply:
It looks the whole thing is not need if we dropped EnableSourcePortBasedLoadBalancing